### PR TITLE
RET-3672: Update to sort listed hearings by datetime for Send a notif…

### PR DESCRIPTION
**Pull request by robin-ieong**
**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x ] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/service/hearings/HearingSelectionService.java [hunk 1]
```diff
@@ -10,37 +11,63 @@
...
 import java.util.ArrayList;
+import java.util.Comparator;
 import java.util.List;
+import java.util.stream.Collectors;
 
 @Service
 public class HearingSelectionService {
 
-    public List<DynamicValueType> getHearingSelection(CaseData caseData, String format) {
+    public static final String HEARING_FORMAT = "%s: %s - %s - %s";
```
Comment (Itay-Kirsh) [inline src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/service/hearings/HearingSelectionService.java:24]: can be private

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/utils/ListedHearingData.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
+package uk.gov.hmcts.ethos.replacement.docmosis.utils;
+
+import lombok.Data;
+import uk.gov.hmcts.et.common.model.bulk.types.DynamicFixedListType;
+
+/**
+ Listed Hearing Data object for sorting purposes
+ */
+@Data
```
Comment (Itay-Kirsh) [inline src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/utils/ListedHearingData.java:9]: @AllArgsConstructor

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/service/hearings/HearingSelectionService.java [hunk 3]
```diff
@@ -10,37 +11,63 @@
...
+    /**
+     * Returns hearing list sorted by datetime only.
+     */
+    public List<DynamicValueType> getHearingSelectionSortedByDateTime(CaseData caseData) {
         List<DynamicValueType> values = new ArrayList<>();
         int index = 1;
 
+        List<ListedHearingData> hearingList = new ArrayList<>();
+
         if (CollectionUtils.isNotEmpty(caseData.getHearingCollection())) {
```
Comment (Itay-Kirsh) [inline src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/service/hearings/HearingSelectionService.java:35]: guard clause this to make the indentation less hard to follow

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/service/hearings/HearingSelectionService.java [hunk 4]
```diff
@@ -10,38 +11,66 @@
...
-                            hearingValue.getHearingType(),
-                            venue,
-                            date);
-                    values.add(DynamicValueType.create(code, label));
-                    index++;
-                }
-            }
+        List<ListedHearingData> sortedHearingList = hearingList.stream()
+                .sorted(Comparator.comparing((ListedHearingData h) -> LocalDateTime.parse(h.getListedDate())))
+                .collect(Collectors.toList());
```
Comment (Itay-Kirsh) [inline src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/service/hearings/HearingSelectionService.java:53]: .toList with java 17.

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/service/hearings/HearingSelectionService.java [hunk 5]
```diff
@@ -10,38 +11,66 @@
...
+    public List<DynamicValueType> getHearingSelectionSortedByDateTime(CaseData caseData) {
         List<DynamicValueType> values = new ArrayList<>();
         int index = 1;
 
-        if (CollectionUtils.isNotEmpty(caseData.getHearingCollection())) {
-            for (HearingTypeItem hearing : caseData.getHearingCollection()) {
-                HearingType hearingValue = hearing.getValue();
-                for (DateListedTypeItem listing : hearingValue.getHearingDateCollection()) {
-                    String code = listing.getId();
+        List<ListedHearingData> hearingList = new ArrayList<>();
```
Comment (Itay-Kirsh) [inline src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/service/hearings/HearingSelectionService.java:33]: Move this just before the first for, not before the guard clause

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/service/hearings/HearingSelectionService.java [hunk 6]
```diff
@@ -10,38 +11,66 @@
...
-                    DynamicFixedListType hearingVenue = hearingValue.getHearingVenue();
+        if (CollectionUtils.isEmpty(caseData.getHearingCollection())) {
+            return null;
+        }
 
-                    String venue = hearingVenue == null ? hearingValue.getHearingVenueScotland() :
-                        hearingVenue.getValue().getLabel();
+        for (HearingTypeItem hearing : caseData.getHearingCollection()) {
+            HearingType hearingValue = hearing.getValue();
+            hearingValue.getHearingDateCollection().stream()
```
Comment (Itay-Kirsh) [inline src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/service/hearings/HearingSelectionService.java:41]: This looks quite complex still

**Review by Itay-Kirsh — CHANGES_REQUESTED** [review]

**Review by Itay-Kirsh — APPROVED** [review]

**Review by JackReevies — APPROVED** [review]
